### PR TITLE
CIMCORE / Canonical JSON

### PR DESCRIPTION
**Pull request by Abhijay**
This adds better hierarchical information, modification information, and constraint history.

The `expandHierarchy` function recursively traverses each Element's `basedOn`, building off of each parent's hierarchical and inheritance information to form a more complete child element.

The `manageValueInheritance` matches child properties with parent properties to determine the original source of inheritance (either the parent itself, or the parent has pre-expanded inheritance information the child can use).

The `manageCardHistory` compiles a history of cardinality changes in a similar fashion and appends that to the child's cardinality.

### CODE REVIEW

**Comment by Abhijay** [comment]
I see the unit tests fail... I'll look into that.

File: lib/expand.js [hunk 1]
```diff
@@ -47,7 +47,7 @@ class Expander {
     const entryDE = this._expanded.dataElements.find('shr.base', 'Entry');
     let entryFields = null;
     if (!entryDE) {
-      logger.warn('Could not find expanded definition of shr.base.Entry. Inheritance calculations will be incomplete.');
+      logger.warn('Could not find expanded definition of shr.base.Entry. Inheritance calculations will be incomplete. ERROR_CODE:TBD');
```
Comment (cmoesel) [inline lib/expand.js]: I assume you'd like to add an error code?

File: lib/expand.js [hunk 2]
```diff
@@ -67,7 +67,7 @@ class Expander {
           }
           const parent = this._expanded.dataElements.findByIdentifier(basedOn);
           if (!parent) {
-            logger.error(`Could not find based on element %s for child element %s.`, basedOn, de.identifier);
+            logger.error(`Could not find based on element %s for child element %s. ERROR_CODE:TBD`, basedOn, de.identifier);
```
Comment (cmoesel) [inline lib/expand.js]: Another TBD error code...

File: lib/expand.js [hunk 3]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+          logger.error('Reference to non-existing base: %s. ERROR_CODE:12002', baseID.fqn);
+          continue;
+        }
+
+      
+        if (base.hierarchy.length == 0 && base.basedOn.length > 0) {
+          this.expandHierarchy(base);
+        }
+
+        var manageValueInheritance = (parent, child) => {
```
Comment (cmoesel) [inline lib/expand.js]: Not a big deal, but `const` is probably more appropriate here.

File: lib/expand.js [hunk 4]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+            if (parent.inheritedFrom) {
+              child.inheritedFrom = parent.inheritedFrom;
+            } else {
+              child.inheritedFrom = base.identifier;
+            };
+          }
+
+          if (child.constraints.length > 0 && child.identifier) {
+            for (const c of child.constraints) {
+              let matchedParentConstraint = parent.constraints.filter(p => p.equals(c)).pop();
```
Comment (cmoesel) [inline lib/expand.js]: We think there will be zero or one of these, right?  In that case, wouldn't [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) be more appropriate?
```js
const matchedParentConstraint = parent.constraints.find(p => p.equals(c));
```
(Also note use of `const` since that variable doesn't change in the scope of the for loop iteration).

File: lib/expand.js [hunk 5]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+        }
+
+        var manageCardHistory = (parent, child) => {
+          if (child.effectiveCard.equals(child.card)) {
+            return;
+          } else {
+            if (parent.constraintsFilter.own.card.constraints.length > 0) {
+              let oldCard = parent.constraintsFilter.own.card.constraints[0].card.clone();
+              oldCard.source = base.identifier;
+              if (parent.card.history.filter(h => h.source == oldCard.source).length == 0) { //if unique
```
Comment (cmoesel) [inline lib/expand.js]: For improved readability, I'd probably use `find`, `some`, or `every`:
```js
if (! parent.card.history.find(h => h.source == oldCard.source)) {
// or
if (! parent.card.history.some(h => h.source == oldCard.source)) {
// or
if (parent.card.history.every(h => h.source != oldCard.source)) {
```

File: lib/expand.js [hunk 6]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+            }
+          } 
+        }
+
+        var manageCardHistory = (parent, child) => {
+          if (child.effectiveCard.equals(child.card)) {
+            return;
+          } else {
+            if (parent.constraintsFilter.own.card.constraints.length > 0) {
+              let oldCard = parent.constraintsFilter.own.card.constraints[0].card.clone();
```
Comment (cmoesel) [inline lib/expand.js]: Again, you can use `const` here.  I won't call this out anymore though...

File: lib/expand.js [hunk 7]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+          if (baseField) {
+            manageValueInheritance(baseField, ef);
+            manageCardHistory(baseField, ef);
+          } else {
+            setCurrentElementAsOriginalModifier(ef);
+          }
+        }
+      
+        element.hierarchy.push(...base.hierarchy);
+        element.hierarchy.push(base);
```
Comment (cmoesel) [inline lib/expand.js]: `push` is a var-arg function, so you should be able to shorten this to one line (if you want).
```js
element.hierarchy.push(...base.hierarchy, base);
```

File: lib/expand.js [hunk 8]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+        var manageValueInheritance = (parent, child) => {
+          if (child.inheritedFrom == null) {
+            if (parent.inheritedFrom) {
+              child.inheritedFrom = parent.inheritedFrom;
+            } else {
+              child.inheritedFrom = base.identifier;
+            };
+          }
+
+          if (child.constraints.length > 0 && child.identifier) {
```
Comment (cmoesel) [inline lib/expand.js]: Why the check on `child.identifier`?  You don't seem to use `child.identifier` below, so it's not a defensive check (as best I can tell).  One thing this check _does_ do is it blocks out `ChoiceValue`s (since they don't have an `identifier`).  Is that intentional?

File: lib/expand.js [hunk 9]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+              };
+            }
+          } 
+        }
+
+        var manageCardHistory = (parent, child) => {
+          if (child.effectiveCard.equals(child.card)) {
+            return;
+          } else {
+            if (parent.constraintsFilter.own.card.constraints.length > 0) {
```
Comment (cmoesel) [inline lib/expand.js]: You can actually use `hasConstraints` here instead...
```js
if (parent.constraintsFilter.own.card.hasConstraints) {
```

File: lib/expand.js [hunk 10]
```diff
@@ -225,6 +227,102 @@ class Expander {
...
+                parent.card.withHistory(oldCard);
+              }
+            } else {
+              let oldCard = parent.card.clone();
+              oldCard.source = base.identifier;
+              child.card.withHistory(oldCard)
+            }
+
+            if (parent.card.history) {
+              child.card.withHistories(parent.card.history);
```
Comment (cmoesel) [inline lib/expand.js]: This adds the parent's histories _after_ the child's histories.  So the most recent history is at index `0` and they get older after that (and the very _original_ card is last in the array).  Just checking that this is the intentional order (as opposed to having index `0` represent the _original_ card).

**Review by cmoesel — CHANGES_REQUESTED** [review]
Overall this looks good, and some nice use of recursion to build up the hierarchies!  I've just sprinkled in a few comments and suggestions.

Also, shr-models 5.3.0 has been released, so you can:
```
yarn upgrade shr-models@^5.3.0
```
and then Travis should be able to run the tests.

File: lib/expand.js (hunk 8 above)
Comment (Abhijay) [reply lib/expand.js]: This was my original intention. I believe my code was originally written in a way that required the matching of the child.identifier to parent.identifier, although, looking at it now, there is no longer any reason to single them out as I've since improved the check to not require that.

**Review by cmoesel — APPROVED** [review]
The latest commits look good, the tests pass, and I've confirmed that this doesn't seem to affect existing exporters.  Let's merge it!

**Comment by cmoesel** [comment]
I've just rebased this on master and squashed some commits to clean up the history.
